src/test/java/seedu/address/model/ModelHistoryManagerTest.java

```diff
@@ -0,0 +1,128 @@
+    private ModelHistoryManager hm;
+    private Participant newP;
+
+    @BeforeEach
+    void beforeEach() throws AlfredException {
+        pList = TypicalParticipants.getTypicalParticipantList();
+        ParticipantList.setLastUsedId(1);
+        mList = TypicalMentors.getTypicalMentorList();
+        MentorList.setLastUsedId(2);
+        tList = TypicalTeams.getTypicalTeamList();
+        TeamList.setLastUsedId(3);
+        hm = new ModelHistoryManager(pList, ParticipantList.getLastUsedId(),
+                                     mList, MentorList.getLastUsedId(),
+                                     tList, TeamList.getLastUsedId());
+        newP = new Participant(new Name("Test Person"),
+                               new Id(PrefixType.P, 123),
+                               new Email("testperson@gmail.com"),
+                               new Phone("93200000"));
+    }
+
+    @Test
+    void updateHistory_changeToEntityList_success() throws AlfredModelHistoryException, AlfredException {
+        pList.add(newP);
+        hm.updateHistory(pList, ParticipantList.getLastUsedId(),
+                         mList, MentorList.getLastUsedId(),
+                         tList, TeamList.getLastUsedId());
+        assertEquals(2, hm.getLengthOfHistory());
+        assertTrue(hm.canUndo());
+    }
+
+    @Test
+    void updateHistory_noEntityChange_success() throws AlfredModelHistoryException {
+        hm.updateHistory(pList, ParticipantList.getLastUsedId(),
+                         mList, MentorList.getLastUsedId(),
+                         tList, TeamList.getLastUsedId());
+        assertEquals(2, hm.getLengthOfHistory());
+        assertTrue(hm.canUndo());
+    }
+
+    @Test
+    void canUndo_initialModelHistoryManager_false() {
+        assertFalse(hm.canUndo());
+    }
+
+    @Test
+    void canUndo_entityChange_true() throws AlfredModelHistoryException {
+        hm.updateHistory(pList, ParticipantList.getLastUsedId(),
+                         mList, MentorList.getLastUsedId(),
+                         tList, TeamList.getLastUsedId());
+        assertTrue(hm.canUndo());
+    }
+
+    @Test
+    void canRedo() {
+        //TODO: Update this in v1.3-1.4
+        assertFalse(hm.canRedo());
+    }
+
+    @Test
+    void undo_testEqualityOfLists_success() throws AlfredException {
+        pList.add(newP);
+        hm.updateHistory(pList, ParticipantList.getLastUsedId(),
+                         mList, MentorList.getLastUsedId(),
+                         tList, TeamList.getLastUsedId());
+        hm.updateHistory(pList, ParticipantList.getLastUsedId(),
+                         mList, MentorList.getLastUsedId(),
+                         tList, TeamList.getLastUsedId());
+
+        assertTrue(hm.canUndo());
+        ModelHistoryRecord hr = hm.undo();
+        ParticipantList historyPList = hr.getParticipantList();
+        assertEquals(pList.getSpecificTypedList(), historyPList.getSpecificTypedList());
+    }
```

Comment:
Hahaha I don't really understand this, so you update hm twice with the same values right? If that is  the case, historyPList will definitely be equal to pList regardless of whether undo method works? Hahaha sorry I don't understand this part:)